composer/datasets/streaming/format.py

```diff
@@ -0,0 +1,286 @@
+]
+
+
+def get_index_basename() -> str:
+    """Get the basename for a streaming dataset index.
+
+    Returns:
+        str: Basename of file.
+    """
+    return 'index.mds'
+
+
+def get_shard_basename(shard: int) -> str:
+    """Get the basename for a streaming dataset shard.
+
+    Args:
+        shard (int): Shard index.
+
+    Returns:
+        str: Basename of file.
+    """
+    return f'{shard:06}.mds'
+
+
+def sample_dict_to_bytes(obj: Dict[str, bytes], keys: List[str]) -> bytes:
+    """Dump a sample dict to bytes, given field names.
+
+    Args:
+        obj (Dict[str, bytes]): The sample dict to encode.
+        keys (list of str): The field names.
+
+    Returns:
+        bytes: The encoded sample bytes.
+    """
+    values = []
+    for key in keys:
+        value = obj[key]
+        values.append(value)
+    sizes = list(map(len, values))
+    sizes = np.array(sizes, np.int64)
+    return sizes.tobytes() + b''.join(values)
+
+
+def bytes_to_sample_dict(data: bytes, keys: List[str]) -> Dict[str, bytes]:
+    """Load a sample dict from bytes and field names.
+
+    Args:
+        data (bytes): The encoded sample data.
+        keys (List[str]): The field names. Must be in the same order as the ``keys`` used when calling :func:`.sample_dict_to_bytes`.
+
+    Returns:
+        Dict[str, bytes]: The decoded sample dict.
+    """
+    num_values = len(keys)
+    sizes = np.frombuffer(data[:num_values * np.int64().nbytes], np.int64)
+    ends = num_values * np.int64().nbytes + sizes.cumsum()
+    begins = ends - sizes
+    values = []
+    for begin, end in zip(begins, ends):
+        value = data[begin:end]
+        values.append(value)
+    return dict(zip(keys, values))
+
+
+def read_array(fp: BufferedIOBase, count: int, dtype: type) -> np.ndarray:
+    """Load the count items from the file handle, advancing its position.
+
+    Args:
+        fp (BufferedIOBase): File handle.
+        count (int): Number of items to read.
+        dtype (type): Item datatype.
+
+    Returns:
+        np.ndarray: The read array.
+    """
+    num_bytes = count * dtype().nbytes
+    data = fp.read(num_bytes)
+    return np.frombuffer(data, dtype)
+
+
+class StreamingDatasetIndex(object):
+    """Streaming Dataset index file, containing all the info about shards and samples.
+
+    The shards are binary buffers with samples concatenated together. All the
+    offset info across the whole dataset is contained in the index file. Workers
+    read this file to calculate how much of which shards their slice is.
+
+    Each sample is a dict of str to bytes. All samples must contain the same
+    dict keys (fields). These strings are stored in the index file for
+    efficiency.
+    """
+
+    def __init__(self, samples_per_shard: NDArray[np.int64], bytes_per_shard: NDArray[np.int64],
+                 bytes_per_sample: NDArray[np.int64], fields: List[str]) -> None:
+        """Initialize with sample statistics.
+
+        Args:
```

Comment:
Can you document these ars in the class-level docstring above? (See other classes in Composer as examples)